### PR TITLE
Skip AppleDouble sidecars during file discovery

### PR DESCRIPTION
**Pull request by nakamuraos**
## Problem

On filesystems without extended attribute support (exFAT, FAT32, and many network shares), macOS cannot store a file's xattrs inline, so it persists them in an AppleDouble sidecar named `._<original>`. The sidecar mirrors the original filename **including the extension**, so `src/app.rs` produces `src/._app.rs`.

`is_rust_file` classifies candidates by extension alone:

```rust
fn is_rust_file(path: &Path) -> bool {
    path.is_file()
        && match path.extension() {
            Some(ext) => match ext.to_str() {
                Some(ext) => ext.eq("rs"),
                None => false,
            },
            None => false,
        }
}
```

`._app.rs` has extension `rs`, so it passes and is handed to `parse_file`. Its contents are binary (AppleDouble magic `0005160700020000`), not UTF-8, so `read_to_string` returns `ErrorKind::InvalidData` and the macro panics:

```
error: custom attribute panicked
 --> src/doc.rs:7:1
  |
7 | #[utoipauto]
  | ^^^^^^^^^^^^
  |
  = help: message: Failed to parse file ./src
```

This makes `#[utoipauto]` unusable on affected checkouts. It is also not something a user can work around durably: the sidecars are regenerated by the OS on every file write, so deleting them fixes exactly one build.

## Why this has been hard to diagnose

`discover_from_file` discarded the underlying `io::Error` and reported the top-level scan root:

```rust
let files = parse_files(&src_path).unwrap_or_else(|_| panic!("Failed to parse file {}", src_path));
```

So the message names `./src`, not the file that actually failed, and it gives no hint that the failure was I/O rather than a syntax error. I suspect this is the root cause behind #54, which reports the identical `Failed to parse file ./src` symptom and was closed without a reproduction. That reporter attributed it to their workspace layout, which may well have been a red herring.

This PR propagates the path and the underlying error, so any future parse failure names the offending file:

```
Failed to parse file ./src: Failed to read ./src/._app.rs: stream did not contain valid UTF-8
```

## Changes

- `is_rust_file` rejects filenames beginning with `._`
- `parse_file` attaches the file path to read errors
- `discover_from_file` includes the error in its panic message
- Regression test covering both `is_rust_file` and `parse_files`

## Note on platform gating

The new check is deliberately **not** behind `#[cfg(target_os = "macos")]`. AppleDouble sidecars are ordinary files that get committed and travel with the repository, so a Linux or CI checkout of a tree authored on such a volume hits exactly the same failure. Gating on the host OS would leave that case broken.

Also note the sidecars are never valid Rust regardless of platform, so skipping them cannot suppress a legitimate source file. The `._` prefix is not usable as a Rust module name.

## Testing

The regression test writes the sidecar bytes directly rather than relying on macOS to generate them, so it reproduces the bug on any platform and will run in CI unchanged:

```rust
#[test]
fn test_apple_double_sidecars_are_not_rust_files() {
    let dir = std::env::temp_dir().join("utoipauto_apple_double_test");
    let _ = fs::remove_dir_all(&dir);
    fs::create_dir_all(&dir).unwrap();

    let source = dir.join("controller.rs");
    fs::write(&source, "pub fn handler() {}\n").unwrap();

    // AppleDouble magic number followed by bytes that are not valid UTF-8.
    let sidecar = dir.join("._controller.rs");
    fs::write(&sidecar, [0x00, 0x05, 0x16, 0x07, 0x00, 0x02, 0xb0, 0xff]).unwrap();

    assert!(is_rust_file(&source));
    assert!(!is_rust_file(&sidecar));

    let parsed = parse_files(&dir).unwrap();
    assert_eq!(parsed.len(), 1);

    fs::remove_dir_all(&dir).unwrap();
}
```

Verified:

- The test fails on `main` without the `is_rust_file` guard (`assertion failed: !is_rust_file(&sidecar)`) and passes with it.
- `cargo test --workspace` passes: 50 tests across 7 test binaries.
- `cargo fmt --all --check` is clean.

## Environment where this was hit

- macOS, repository checked out on an exFAT volume (`/dev/disk6s2 on /Volumes/data (exfat, local, ...)`)
- `utoipauto` 0.2.0 with `utoipa` 5.5.0, reproduced against `main` (0.3.0-alpha.2)
- 13 sidecars were present under `src/`; each edit to a source file regenerated its sidecar

### CODE REVIEW

File: utoipauto-core/src/file_utils.rs [hunk 1]
```diff
@@ -56,6 +58,17 @@ fn is_rust_file(path: &Path) -> bool {
...
 
+/// Filesystems without extended attribute support (exFAT, FAT32, and many
+/// network shares) make macOS persist a file's metadata in an AppleDouble
+/// sidecar named `._<original>`. The sidecar mirrors the original name,
+/// extension included, so it is indistinguishable from a Rust source file by
+/// extension alone while holding a binary payload that is not valid UTF-8.
+fn is_apple_double(path: &Path) -> bool {
+    path.file_name()
+        .and_then(|name| name.to_str())
+        .is_some_and(|name| name.starts_with("._"))
```
Comment (DenuxPlays) [inline utoipauto-core/src/file_utils.rs]: We should only do this if we actually are on macos.

**Review by DenuxPlays — CHANGES_REQUESTED** [review]
This is a very specific issue and I think this should be fixed by apple.
It seems that these kind of things break a lot of stuff but apple seems to like their weird random sys files

**Comment by nakamuraos** [comment]
Thanks for taking a look. No worries - I'll close this.

One note in case it helps someone who runs into the same issue later.

The `._*` files aren't just Finder clutter that macOS occasionally leaves behind. They're AppleDouble sidecar files. When a file has extended attributes but lives on a filesystem that can't store them inline (such as exFAT, FAT32, or many network filesystems), macOS stores that metadata in a separate `._*` file instead. That's expected behavior rather than a bug, so it's unlikely to change.

This also isn't only a macOS problem. Once those sidecar files are committed, they're just regular files in the repository, so Linux users and CI environments will hit the same issue.

If anyone ends up here after seeing `Failed to parse file ./src` and can't move their checkout off exFAT, a small `build.rs` works around it nicely. Manually deleting the sidecar files isn't enough because macOS recreates them whenever the original files are modified. Running the cleanup from `build.rs` ensures they're removed before the crate is compiled and before the macro scans the source tree.

`(project)/build.rs`:

```rust
use std::fs;
use std::path::Path;

fn main() {
    println!("cargo::rerun-if-changed=.");
    remove_apple_double_files(Path::new("."));
}

fn remove_apple_double_files(dir: &Path) {
    let Ok(entries) = fs::read_dir(dir) else {
        return;
    };

    for entry in entries.flatten() {
        let path = entry.path();
        let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
            continue;
        };

        if path.is_dir() {
            if name != "target" && name != ".git" {
                remove_apple_double_files(&path);
            }
            continue;
        }

        if name.starts_with("._") {
            let _ = fs::remove_file(&path);
        }
    }
}
```

Cargo automatically picks up `build.rs` from the package root.

One other thing: the second half of this PR wasn't really specific to Apple.

Right now `discover_from_file` discards the underlying `io::Error` and only reports the scan root:

```rust
let files = parse_files(&src_path)
    .unwrap_or_else(|_| panic!("Failed to parse file {}", src_path));
```

That means every read or parse failure ends up looking like:

```text
Failed to parse file ./src
```

which doesn't indicate which file actually failed.

If the original error is propagated instead, the message becomes something like:

```text
Failed to parse file ./src: Failed to read ./src/._app.rs: stream did not contain valid UTF-8
```

It's only a small change, but it makes every parse failure much easier to diagnose, regardless of the underlying cause. I suspect that's also why #54 was difficult to investigate - the original reporter had no way to tell which file was triggering the error.

If that's something you'd be interested in, I'm happy to open a separate PR with just that change. If not, no worries at all.

**Comment by DenuxPlays** [comment]
No you do not need to close it.
If it does not work on macos I am all for it to fix it.

We just should not check this on every os imo

**Comment by nakamuraos** [comment]
> We just should not check this on every os imo

What happens if a user accidentally commits `._*` files to Git and then checks out the repository on another OS?

Since AppleDouble files are just regular files outside macOS, could we improve this by supporting include/exclude patterns, for example:

```
#[utoipauto(exclude = "**/._*")]
```

What do you think? Should this stay as a macOS-specific fix, or should I be open to a separate PR adding include/exclude pattern support, for example to ignore files like `._*`?

**Comment by DenuxPlays** [comment]
First of all thanks for the improved error messages (I forgot that in my review statement sorry for that).

Well that's a discussion point.
I see where someone would accidental commit these files but also they can be ignored easily.

However I do like the idea of exclusion-pattern support.
I think we could even apply `**/._*` as the default pattern since I doubt that someone is affected in a negative way.
But there has to be a way to disable it, so no files are getting excluded.

**Comment by nakamuraos** [comment]
Thanks @DenuxPlays, that makes sense. I've pushed an implementation for the `exclude` option:

```rust
#[utoipauto(paths = "./src", exclude = "**/generated/**, **/._*")]
```

`exclude` accepts comma-separated glob patterns and defaults to `**/._*`, so AppleDouble files are ignored automatically. A custom value replaces the default, while `exclude = ""` disables filtering entirely.

I also kept the improved parse error message, added tests for the default, empty, and custom exclude cases, and documented the option in the README.

File: utoipauto-core/src/file_utils.rs [hunk 2]
```diff
@@ -36,15 +43,28 @@ pub fn parse_files<T: Into<PathBuf>>(path: T) -> Result<Vec<(String, syn::File)>
...
         }
     }
     Ok(files)
 }
 
+/// Build a matcher from comma-separated glob patterns. Blank entries are
+/// ignored, so an empty string yields a set that matches nothing.
+pub fn build_exclude_set(patterns: &str) -> GlobSet {
+    let mut builder = GlobSetBuilder::new();
+    for pattern in patterns.split(',').map(str::trim).filter(|p| !p.is_empty()) {
```
Comment (DenuxPlays) [inline utoipauto-core/src/file_utils.rs]: This would not support alternate globs right?
Example:
`glob1.{txt,text},glob2*`

Not sure if there is a easy way to fix it but we should atleast add a test & document this.

**Review by DenuxPlays — CHANGES_REQUESTED** [review]
Looks great just for one thing.

And:

It's been a while working with macros but is a syntax like this possible:
```rs
#[utoipauto(exclude = ["**/._*",  "**/generated/**"])]
```
instead of:
```rs
#[utoipauto(exclude = "**/._*, **/generated/**")]
```

Please let me know what you think about this approach.

**Comment by nakamuraos** [comment]
I initially considered using an array, as you suggested. However, I noticed that `paths` uses a comma-separated string, so I implemented it the same way to keep the value format consistent across attributes.

Personally, I think using an array would make things simpler and easier to maintain in the long run.

While looking into this, I also realized that `paths` isn't just a plain string, it actually uses a small DSL implemented in [utoipauto-core/src/string_utils.rs:28](https://github.com/ProbablyClem/utoipauto/blob/32a731f462d0a00355f3b0d9b11617ca3bd878ab/utoipauto-core/src/string_utils.rs#L46)

**Comment by DenuxPlays** [comment]
Oh yeah we should be consistent. nice catch

Then it's fine (except we need to document this that globs contains `,` are not supported).
We should refactor the macro to use arrays and find a solution for the DSL sometime.

**Comment by nakamuraos** [comment]
Thanks. I've updated the code to use the array.

**Review by DenuxPlays — APPROVED** [review]

**Comment by DenuxPlays** [comment]
Needs: https://github.com/ProbablyClem/utoipauto/pull/57

**Comment by DenuxPlays** [comment]
@nakamuraos 
Could you rebase this pr on main to see if the test now pass

**Comment by DenuxPlays** [comment]
Oh yeah we need to rust 1.88.0 for this.
Not a problem I will prepare this this will inline our msrv with utoipa's
https://github.com/juhaku/utoipa/pull/1562

**Comment by DenuxPlays** [comment]
I've updated to rust 1.88.0 on main

You may need to run clippy and cargo fmt again

**Comment by nakamuraos** [comment]
Thank you, @DenuxPlays
I've run clippy and fmt and fixed the issues.

**Comment by DenuxPlays** [comment]
Nice thank you for your work
I will make a release later
